File: KCD/HMStrokeTextFieldCell.m
//
//  HMStrokeTextFieldCell.m
//  KCD
//
//  Created by Hori,Masaki on 2014/04/19.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

#import "HMStrokeTextFieldCell.h"

const CGFloat boarderWidth = 2.0;

@interface HMStrokeTextFieldCell ()
@property (strong, nonatomic) NSLayoutManager *layoutManager;
@property (strong, nonatomic) NSTextContainer *textContainer;
@end

@implementation HMStrokeTextFieldCell

- (id)initWithCoder:(NSCoder *)aDecoder
{
	self = [super initWithCoder:aDecoder];
	if(self) {
		_layoutManager = [NSLayoutManager new];
		_textContainer = [NSTextContainer new];
		[self.layoutManager addTextContainer:self.textContainer];
	}
	return self;
}

- (void)drawInteriorWithFrame:(NSRect)cellFrame inView:(NSView *)controlView
{
	NSAttributedString *attributedString = self.attributedStringValue;
	NSDictionary *attribute = [attributedString attributesAtIndex:0 effectiveRange:NULL];
	NSColor *forgroundColor = [attribute objectForKey:NSForegroundColorAttributeName];
	if(!forgroundColor) return;
	if([forgroundColor isEqual:[NSColor controlTextColor]]) {
		[super drawInteriorWithFrame:cellFrame inView:controlView];
		return;
	}
//
//	NSTextStorage *textStorage = [[NSTextStorage alloc] initWithString:attributedString.string attributes:attribute];
//	[textStorage addLayoutManager:self.layoutManager];
//	
//	NSRange range = [self.layoutManager glyphRangeForTextContainer:self.textContainer];
//	NSGlyph glyph[range.length];
//	NSUInteger glyphLength = [self.layoutManager getGlyphs:glyph range:range];
//	
//	NSFont *font = [attribute objectForKey:NSFontAttributeName];
//	NSPoint point = {boarderWidth,0};
//	point.y -= [font descender];
//	if([controlView isFlipped]) {
//		point.y -= [controlView frame].size.height;
//	}
//	
//	NSBezierPath *path = [NSBezierPath new];
//	[path moveToPoint:point];
//	[path appendBezierPathWithGlyphs:glyph count:glyphLength inFont:font];
//	[path setLineWidth:boarderWidth];
//	[path setLineJoinStyle:NSRoundLineJoinStyle];
//	if([controlView isFlipped]) {
//		NSAffineTransform *affineTransform = [NSAffineTransform transform];
//		[affineTransform scaleXBy:1 yBy:-1];
//		[path transformUsingAffineTransform:affineTransform];
//	}
//	
//	[[NSColor blackColor] set];
//	[path stroke];
//	
//	[forgroundColor set];
//	[path fill];
	
	NSMutableDictionary *newAttr = [attribute mutableCopy];
	newAttr[NSStrokeColorAttributeName] = [NSColor blackColor];
	newAttr[NSStrokeWidthAttributeName] = @(-4);
	
	[attributedString.string drawInRect:cellFrame withAttributes:newAttr];
	
}
@end

File: KCD/HMStrokeTextFieldCell.swift
//
//  HMStrokeTextFieldCell.swift
//  KCD
//
//  Created by Hori,Masaki on 2014/12/28.
//  Copyright (c) 2014年 Hori,Masaki. All rights reserved.
//

import Cocoa

class HMStrokeTextFieldCell: NSTextFieldCell
{
	let borderWidth: CGFloat = 2.0
	let layoutManager: NSLayoutManager = NSLayoutManager()
	lazy var textContainer: NSTextContainer = {
		let aTextContainer = NSTextContainer()
		self.layoutManager.addTextContainer(aTextContainer)
		return aTextContainer
	}()
	
	override func drawInteriorWithFrame(cellFrame: NSRect, inView controlView: NSView) {
		let attribute = attributedStringValue.attributesAtIndex(0, effectiveRange: nil)
		let forgroundColor = attribute[NSForegroundColorAttributeName] as? NSColor
		if forgroundColor == nil {
			return
		}
		if forgroundColor == NSColor.controlTextColor() {
			super.drawInteriorWithFrame(cellFrame, inView: controlView)
			return
		}
		
		let textStorage = NSTextStorage(string: attributedStringValue.string, attributes: attribute)
		textStorage.addLayoutManager(layoutManager)
		
		let range: NSRange = layoutManager.glyphRangeForTextContainer(textContainer)
		var glyph = [NSGlyph](count: range.length + 1, repeatedValue:0)
		let glyphLength = layoutManager.getGlyphs(&glyph, range: range)
		
		let font = attribute[NSFontAttributeName] as NSFont
		var point = NSPoint(x: borderWidth, y: -font.descender)
		if controlView.flipped {
			point.y -= controlView.frame.size.height
		}
		
		let path = NSBezierPath()
		path.moveToPoint(point)
		path.appendBezierPathWithGlyphs(&glyph, count: glyphLength, inFont: font)
		path.lineWidth = borderWidth
		path.lineJoinStyle = .RoundLineJoinStyle
		if controlView.flipped {
			let affineTransform = NSAffineTransform()
			affineTransform.scaleXBy(1, yBy: -1)
			path.transformUsingAffineTransform(affineTransform)
		}
		
		NSColor.blackColor().set()
		path.stroke()
		
		forgroundColor!.set()
		path.fill()
	}
}
